### PR TITLE
GODRIVER-1808 Fix BSON unmarshaling into an interface containing a concrete value.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-1808

## Summary
~~This PR fixes BSON's unmarshaling behavior for interfaces by passing `reflect.Value` instead of `reflect.Type`, so the actual value type contained in the interface can be used.~~

- Merge `decodeTypeOrValue` into `decodeTypeOrValueWithInfo`
- Simplify the signature of `decodeTypeOrValueWithInfo`
- Support unmarshaling BSON into interface containing concrete value
- Remove unused functions in "bson/bsoncodec/default_value_decoders.go"

## Background & Motivation
This PR makes unmarshaling behavior closer to the `JSON` package.

### CODE REVIEW

File: bson/bsoncodec/default_value_decoders.go [hunk 1]
```diff
@@ -1666,11 +1666,17 @@ func (dvd DefaultValueDecoders) decodeDefault(dc DecodeContext, vr bsonrw.ValueR
 
 	eType := val.Type().Elem()
 
-	decoder, err := dc.LookupDecoder(eType)
-	if err != nil {
-		return nil, err
+	var vDecoder *ValueDecoder
+	var tDecoder *typeDecoder
+	if !(eType.Kind() == reflect.Interface && val.Len() > 0) {
```
Comment (qingyang-hu) [inline bson/bsoncodec/default_value_decoders.go:1671]: Delay the decoder retrieval for interface slice/array so each element will be decoded based on its type.

**Review by matthewdale — CHANGES_REQUESTED** [review]
We should update the implementation to remove the added calls to `reflect.New`, which seems to increase memory allocs in some cases. Is there a way to only pass a `reflect.Type` when we only have type information, and pass a `reflect.Value` when we have value information?

I noticed `BenchmarkUnmarshal` reports about 30% more allocations with this PR. Keep in mind that `BenchmarkUnmarshal` actually only benchmarks unmarhshaling into a `map`. `BenchmarkCodeUnmarshal`, which unmarshals into a struct, shows no difference, so it seems to be isolated to `map` unmarshaling.
```
❯ benchstat old.txt new.txt                                                                                                       
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/bson
                                    │   old.txt   │               new.txt                │
                                    │   sec/op    │    sec/op     vs base                │
Unmarshal/simple_struct/BSON-10       3.423µ ± 0%   3.961µ ± 19%  +15.72% (p=0.000 n=10)
Unmarshal/nested_struct/BSON-10       8.605µ ± 1%   9.258µ ±  0%   +7.59% (p=0.000 n=10)
Unmarshal/deep_bson.json.gz/BSON-10   41.98µ ± 0%   46.45µ ±  1%  +10.66% (p=0.000 n=10)
Unmarshal/flat_bson.json.gz/BSON-10   41.88µ ± 1%   47.73µ ±  0%  +13.97% (p=0.000 n=10)
Unmarshal/full_bson.json.gz/BSON-10   49.61µ ± 0%   55.90µ ±  0%  +12.67% (p=0.000 n=10)
geomean                               19.14µ        21.45µ        +12.09%

                                    │   old.txt    │               new.txt                │
                                    │     B/op     │     B/op      vs base                │
Unmarshal/simple_struct/BSON-10       1.062Ki ± 0%   1.375Ki ± 0%  +29.41% (p=0.000 n=10)
Unmarshal/nested_struct/BSON-10       7.927Ki ± 0%   8.411Ki ± 0%   +6.11% (p=0.000 n=10)
Unmarshal/deep_bson.json.gz/BSON-10   30.15Ki ± 0%   33.12Ki ± 0%   +9.85% (p=0.000 n=10)
Unmarshal/flat_bson.json.gz/BSON-10   13.11Ki ± 0%   16.79Ki ± 0%  +28.09% (p=0.000 n=10)
Unmarshal/full_bson.json.gz/BSON-10   19.93Ki ± 0%   23.57Ki ± 0%  +18.27% (p=0.000 n=10)
geomean                               9.212Ki        10.87Ki       +17.97%

                                    │  old.txt   │               new.txt               │
                                    │ allocs/op  │  allocs/op   vs base                │
Unmarshal/simple_struct/BSON-10       62.00 ± 0%    86.00 ± 0%  +38.71% (p=0.000 n=10)
Unmarshal/nested_struct/BSON-10       143.0 ± 0%    178.0 ± 0%  +24.48% (p=0.000 n=10)
Unmarshal/deep_bson.json.gz/BSON-10   822.0 ± 0%   1012.0 ± 0%  +23.11% (p=0.000 n=10)
Unmarshal/flat_bson.json.gz/BSON-10   740.0 ± 0%   1030.0 ± 0%  +39.19% (p=0.000 n=10)
Unmarshal/full_bson.json.gz/BSON-10   797.0 ± 0%   1086.0 ± 0%  +36.26% (p=0.000 n=10)
geomean                               336.2         444.4       +32.16%
```

Here's a graph view of a memory profile of running `BenchmarkUnmarshal` on the previous commit, filtered to only `reflect` calls:
Screenshot 2024-04-19 at 7 06 21 PM
And here's the same profile of the PR commit, filtered to only `reflect` calls:
Screenshot 2024-04-19 at 7 06 30 PM

The allocs from `reflect.New` grew from 0.28GB to 1GB, which seems to be the biggest change between the two profiles.

**Review by prestonvasquez — APPROVED** [review]
LGTM, with updates proposed by Matt.

File: bson/decoder_test.go
Comment (matthewdale) [inline bson/decoder_test.go:1]: We should add a test or a case to an existing test that asserts that this behavior works when calling `bson.Unmarshal` in addition to `TestDecodingInterfaces`.

**Review by matthewdale — CHANGES_REQUESTED** [review]
Looks like the additional unmarshaling allocations are fixed! One more request for a test, but everything else looks good!

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

**Review by prestonvasquez — APPROVED** [review]

**Pull request merged**